edgePy/DGEList.py

```diff
@@ -292,17 +292,40 @@ def rpkm(
         # self = self.cpm(log=log, prior_count=prior_count)
 
     def tpm(
-        self, transcripts: Mapping, log: bool = False, prior_count: float = PRIOR_COUNT
-    ) -> None:
-        """Return the DGEList normalized to reads per kilobase of transcript
-        length.
+        self,
+        gene_lengths: np.ndarray,
+        log: bool = False,
+        prior_count: float = PRIOR_COUNT,
+        mean_fragment_lengths: np.ndarray = None,
+    ) -> np.ndarray:
+        """Normalize the DGEList to transcripts per million
+
+        Args:
+            gene_lengths: vector of length for each gene in counts table
+            log: store log outputs
+            prior_count:
+            mean_fragment_lengths: vector of mean fragment length for each sample in counts table (optional)
 
         """
-        raise NotImplementedError
 
-        # TODO: Implement here
+        # based on https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/ and
+        # https://gist.github.com/slowkow/c6ab0348747f86e2748b
```

Comment:
Use the formulation or derivation as published and cite that:

```
Wagner, et al. Measurement of mRNA abundance using RNA-seq data:
RPKM measure is inconsistent among samples. Theory Biosci. 24 July 2012.
doi:10.1007/s12064-012-0162-3
```

Downside of citing other algorithm implementations (from source or gists) is that you inherit their assumptions (and bugs) even if you don't intend to.

Reply:
Wanted to do that so I read the paper, but I was not 100% sure about their formulae as it includes one additional factor which the implementations apparently don't use
From publication:
![image](https://user-images.githubusercontent.com/2714540/46858048-b9fa8d00-ce0a-11e8-8b26-f75c701b0cb1.png)
Are you familiar with it and could elucidate it?

Reply:
Not terribly familiar, sorry.

I can only really advise on topics of API design, overall engineering, and scientific Python in general.

Maybe ping the Slack channel? I would honestly start with examining the implementation as exactly written in the source of `edgeR` to see how they compute.

Reply:
Also! Formulae in docstrings corresponding to the algorithm as implemented would be amazing. If you know a bit of JS and want to add MathJax support for our Sphinx compiled docs I would be 👍 👍 👍 

http://www.sphinx-doc.org/en/master/usage/extensions/math.html

Reply:
lmk if you need help with Sphinx, I use it (too frequently).

Reply:
I'll ask in Slack and check the edgeR source to verify. I'll have a look into the sphinx docs tomorrow or so.